diff -ruN libmseed/libmseed.h libmseed_mod/libmseed.h
--- libmseed/libmseed.h	2019-06-19 10:59:29.000000000 +0200
+++ libmseed_mod/libmseed.h	2021-04-08 13:38:03.819403113 +0200
@@ -511,6 +511,7 @@
   double          samprate;          /* Nominal sample rate (Hz) */
   int64_t         samplecnt;         /* Number of samples in trace coverage */
   void           *datasamples;       /* Data samples, 'numsamples' of type 'sampletype' */
+  size_t          bufsize;
   int64_t         numsamples;        /* Number of data samples in datasamples */
   char            sampletype;        /* Sample type code: a, i, f, d */
   void           *prvtptr;           /* Private pointer for general use, unused by libmseed */
diff -ruN libmseed/Makefile.win libmseed_mod/Makefile.win
--- libmseed/Makefile.win	2019-06-19 10:59:29.000000000 +0200
+++ libmseed_mod/Makefile.win	2021-04-07 16:22:04.091415681 +0200
@@ -5,9 +5,10 @@
 NODEBUG=1
 
 INCS = -I.
-OPTS = -D_CRT_SECURE_NO_WARNINGS
+OPTS = -D_CRT_SECURE_NO_WARNINGS -DWIN32
 LIB = libmseed.lib
 DLL = libmseed.dll
+CFLAGS = /Ox
 
 OBJS=	fileutils.obj	\
 	genutils.obj	\
diff -ruN libmseed/traceutils.c libmseed_mod/traceutils.c
--- libmseed/traceutils.c	2019-06-19 10:59:29.000000000 +0200
+++ libmseed_mod/traceutils.c	2021-04-08 13:47:55.097138422 +0200
@@ -437,6 +437,7 @@
 mst_addmsr (MSTrace *mst, MSRecord *msr, flag whence)
 {
   int samplesize = 0;
+  size_t min_size;
 
   if (!mst || !msr)
     return -1;
@@ -465,13 +466,17 @@
       return -1;
     }
 
-    mst->datasamples = realloc (mst->datasamples,
-                                (size_t) (mst->numsamples * samplesize + msr->numsamples * samplesize));
+    min_size = (size_t) (mst->numsamples * samplesize + msr->numsamples * samplesize);
 
-    if (mst->datasamples == NULL)
-    {
-      ms_log (2, "mst_addmsr(): Cannot allocate memory\n");
-      return -1;
+    if (min_size > mst->bufsize) {
+        mst->bufsize = min_size*2;
+        mst->datasamples = realloc (mst->datasamples, mst->bufsize);
+
+        if (mst->datasamples == NULL)
+        {
+          ms_log (2, "mst_addmsr(): Cannot allocate memory\n");
+          return -1;
+        }
     }
   }
 
